pkg/geoipupdate/database/http_reader.go

```
@@ -0,0 +1,145 @@
+package database
+
+import (
+	"compress/gzip"
+	"fmt"
+	"github.com/maxmind/geoipupdate/pkg/geoipupdate"
+	"github.com/pkg/errors"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+//HTTPDatabaseReader is a database.Reader that uses an HTTP client to retrieve the database data
+type HTTPDatabaseReader struct {
+	client            *http.Client
+	url               string
+	licenseKey        string
+	accountID         int
+	preserveFileTimes bool
+	verbose           bool
+}
+
+func NewHTTPDatabaseReader(client *http.Client, config *geoipupdate.Config) Reader {
+	return &HTTPDatabaseReader{
+		client:            client,
+		url:               config.URL,
+		licenseKey:        config.LicenseKey,
+		accountID:         config.AccountID,
+		preserveFileTimes: config.PreserveFileTimes,
+		verbose:           config.Verbose,
+	}
+}
+
+//Get retrieves the data for a given editionID using an HTTP client to MaxMind, writes it to database.Writer,
+// and validates the associated hash before committing
+func (reader *HTTPDatabaseReader) Get(destination Writer, editionID string) error {
+	defer func() {
+		if err := destination.Close(); err != nil {
+			log.Println(err)
```

Perhaps we could make this fatal?

---

For both this and GZip reader, I would hedge away from Fatals.  Given that all the writers are distinct operations, an error closing out one of them doesn't mean we should abort the entire operation.  If this is the behavior you're looking for though, I'm recommend making them panics instead so all defer calls throughout the stack have an opportunity to execute and clean up as many resources on the way to exiting that it can.

---

Yeah, my concern is that we could ignore the errors if we only log them. One of the drawbacks of defer! I agree that exiting isn't great. Let's leave it.